### PR TITLE
Documentation: Basic info about projections

### PR DESCRIPTION
**Pull request by morozovsk**
Basic info about projections based on RFC https://github.com/ClickHouse/ClickHouse/issues/14730

I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Documentation (changelog entry is not required)

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

...

Detailed description / Documentation draft:

...

By adding documentation, you'll allow users to try your new feature immediately, not when someone else will have time to document it later. Documentation is necessary for all features that affect user experience in any way. You can add brief documentation draft above, or add documentation right into your patch as Markdown files in [docs](https://github.com/ClickHouse/ClickHouse/tree/master/docs) folder.

If you are doing this for the first time, it's recommended to read the lightweight [Contributing to ClickHouse Documentation](https://github.com/ClickHouse/ClickHouse/tree/master/docs/README.md) guide first.

Information about CI checks: https://clickhouse.tech/docs/en/development/continuous-integration/

### CODE REVIEW

File: docs/en/engines/table-engines/mergetree-family/mergetree.md [hunk 1]
```diff
@@ -39,7 +39,10 @@ CREATE TABLE [IF NOT EXISTS] [db.]table_name [ON CLUSTER cluster]
     name2 [type2] [DEFAULT|MATERIALIZED|ALIAS expr2] [TTL expr2],
     ...
     INDEX index_name1 expr1 TYPE type1(...) GRANULARITY value1,
-    INDEX index_name2 expr2 TYPE type2(...) GRANULARITY value2
+    INDEX index_name2 expr2 TYPE type2(...) GRANULARITY value2,
+    ...
+    PROJECTION projection_name_1 (SELECT <COLUMN LIST EXPR> [WHERE] [GROUP BY] [ORDER BY]),
```
Comment (amosbird) [inline docs/en/engines/table-engines/mergetree-family/mergetree.md:44]: Currently `[WHERE]` is not supported. We should remove it for now and add it back afterwards.

File: docs/en/engines/table-engines/mergetree-family/mergetree.md [hunk 2]
```diff
@@ -385,6 +388,26 @@ Functions with a constant argument that is less than ngram size can’t be used
...
+A projection query is what defines a projection. It has the following grammar:
+
+`SELECT <COLUMN LIST EXPR> [WHERE] [GROUP BY] [ORDER BY]`
+
+It implicitly selects data from the parent table.
+
+#### Storage {#projection-storage}
+Projections are stored inside the part directory. It's similar to an index but contains an subdirectory which stores an anonymous MergeTree table's part. The table is induced by the definition query of the projection. If there is a GROUP BY clause, the underlying storage engine becomes AggregatedMergeTree, and all aggregate functions are converted to either AggregateFunction or SimpleAggregateFunction. If there is an ORDER BY clause, the MergeTree table will use it as its primary key expression. During the merge process, the projection part will be merged via its storage's merge routine. The checksum of the parent table's part will combine the projection's part. Other maintenance jobs are similar to skip indices.
+
+#### Query Routing {#projection-query-routing}
```
Comment (amosbird) [inline docs/en/engines/table-engines/mergetree-family/mergetree.md:404]: #### Query Analysis {#projection-query-analysis}

1. Check if the projection can be used to answer the given query, that is, it generates the same answer as querying the base table.

2. Select the best feasible match, which contains the least granules to read.

3. The query pipeline which uses projections will be different from the one that uses the original parts. if the projection is absent in some parts, we can add the pipeline to "project" it on the fly.

File: docs/en/engines/table-engines/mergetree-family/mergetree.md [hunk 3]
```diff
@@ -385,6 +388,26 @@ Functions with a constant argument that is less than ngram size can’t be used
...
+
+#### Query {#projection-query}
+A projection query is what defines a projection. It has the following grammar:
+
+`SELECT <COLUMN LIST EXPR> [WHERE] [GROUP BY] [ORDER BY]`
+
+It implicitly selects data from the parent table.
+
+#### Storage {#projection-storage}
+Projections are stored inside the part directory. It's similar to an index but contains an subdirectory which stores an anonymous MergeTree table's part. The table is induced by the definition query of the projection. If there is a GROUP BY clause, the underlying storage engine becomes AggregatedMergeTree, and all aggregate functions are converted to either AggregateFunction or SimpleAggregateFunction. If there is an ORDER BY clause, the MergeTree table will use it as its primary key expression. During the merge process, the projection part will be merged via its storage's merge routine. The checksum of the parent table's part will combine the projection's part. Other maintenance jobs are similar to skip indices.
```
Comment (amosbird) [inline docs/en/engines/table-engines/mergetree-family/mergetree.md:402]: SimpleAggregateFunction is not used for now.

File: docs/en/engines/table-engines/mergetree-family/mergetree.md [hunk 4]
```diff
@@ -385,6 +388,26 @@ Functions with a constant argument that is less than ngram size can’t be used
     -   `s != 1`
     -   `NOT startsWith(s, 'test')`
 
+### Projections {#projections}
+Projections are like materialized views, but defined in part-level. It provides consistency guarantees along with automatic usage in queries.
+
+#### Query {#projection-query}
+A projection query is what defines a projection. It has the following grammar:
+
+`SELECT <COLUMN LIST EXPR> [WHERE] [GROUP BY] [ORDER BY]`
```
Comment (amosbird) [inline docs/en/engines/table-engines/mergetree-family/mergetree.md:397]: no [WHERE] for now

**Review by morozovsk — COMMENTED** [review]
fixed. thanks for review and for projections

**Comment by json-lrj** [comment]
1. Check if the projection can be used to answer the given query, that is, it generates the same answer as querying the base table.
it seems not correct in not append only mergetree. when rows with same order by key do special 'merges', the projection lost consistent with base table.Do not overclaim for common users.

**Comment by amosbird** [comment]
> 1. Check if the projection can be used to answer the given query, that is, it generates the same answer as querying the base table.
>    it seems not correct in not append only mergetree. when rows with same order by key do special 'merges', the projection lost consistent with base table.Do not overclaim for common users.

Yeah. It deserves a special NOTE for specialized MergeTree tables. Actually I'll just disable creating projections for these engines until it's properly implemented.

File: docs/en/engines/table-engines/mergetree-family/mergetree.md [hunk 5]
```diff
@@ -385,6 +388,24 @@ Functions with a constant argument that is less than ngram size can’t be used
...
+
+#### Query {#projection-query}
+A projection query is what defines a projection. It has the following grammar:
+
+`SELECT <COLUMN LIST EXPR> [GROUP BY] [ORDER BY]`
+
+It implicitly selects data from the parent table.
+
+#### Storage {#projection-storage}
+Projections are stored inside the part directory. It's similar to an index but contains an subdirectory which stores an anonymous MergeTree table's part. The table is induced by the definition query of the projection. If there is a GROUP BY clause, the underlying storage engine becomes AggregatedMergeTree, and all aggregate functions are converted to AggregateFunction. If there is an ORDER BY clause, the MergeTree table will use it as its primary key expression. During the merge process, the projection part will be merged via its storage's merge routine. The checksum of the parent table's part will combine the projection's part. Other maintenance jobs are similar to skip indices.
```
Comment (mathalex) [inline docs/en/engines/table-engines/mergetree-family/mergetree.md:402]:
```suggestion
Projections are stored inside the part directory. It's similar to an index but contains a subdirectory that stores an anonymous MergeTree table's part. The table is induced by the definition query of the projection. If there is a GROUP BY clause, the underlying storage engine becomes AggregatedMergeTree, and all aggregate functions are converted to AggregateFunction. If there is an ORDER BY clause, the MergeTree table will use it as its primary key expression. During the merge process, the projection part will be merged via its storage's merge routine. The checksum of the parent table's part will combine the projection's part. Other maintenance jobs are similar to skip indices.
```
Here we have an essential clause.

File: docs/en/engines/table-engines/mergetree-family/mergetree.md [hunk 6]
```diff
@@ -385,6 +388,24 @@ Functions with a constant argument that is less than ngram size can’t be used
     -   `s != 1`
     -   `NOT startsWith(s, 'test')`
 
+### Projections {#projections}
+Projections are like materialized views, but defined in part-level. It provides consistency guarantees along with automatic usage in queries.
```
Comment (mathalex) [inline docs/en/engines/table-engines/mergetree-family/mergetree.md:392]:
```suggestion
Projections are like materialized views but defined in part-level. It provides consistency guarantees along with automatic usage in queries.
```
Clauses are not independent, so the comma is not needed here.

**Review by mathalex — APPROVED** [review]

**Comment by olgarev** [comment]
Internal documentation ticket DOCSUP-12894

**Pull request merged**